### PR TITLE
[WIP] Normalizing flows

### PR DESCRIPTION
**Pull request by ncfrey**
This PR is an initial pass on an API for normalizing flows (NFs) in DeepChem.

### Summary
- `NormalizingFlow` abstract base class for creating NFs with TensorFlow Probability.
- `NormalizingFlowModel` model class that wraps a base distribution and a NF to allow for training and inference.
- `NormalizingFlowLayer` abstract base class for implementing new NFs.

### To-do
- Implement new loss functions appropriate for NFs
- Implement NF models for specific applications: operating on molecular graphs, etc.
- Write example notebook

### CODE REVIEW

File: deepchem/models/normalizing_flows.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+    with tf.GradientTape() as tape:
+      dummy_labels = np.ones(len(x))
+      batch_loss = loss(x, dummy_labels)
+      grads = tape.gradient(batch_loss, self.trainable_variables)
+      optimizer.apply_gradients(zip(grads, self.trainable_variables))
+    return batch_loss
+
+
+class NormalizingFlowModel(NormalizingFlow):
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:76]: Would it be possible for NormalizingFlowModel to be a subclass of `dc.models.Model`?

File: deepchem/models/normalizing_flows.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+    ----------
+    base_distribution : tfd.Distribution
+      Probability distribution to be transformed.
+      Typically an N dimensional multivariate Gaussian.
+    flow_layers : Sequence[tfb.Bijector]
+      An iterable of bijectors that comprise the flow.
+    optimizer: Optional[tf.keras.optimizers.Optimizer]
+      An instance of Optimizer.
+    loss: Optional[Any]
+      Loss function, e.g. an instance of dc.models.losses.Loss.
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:123]: Is there anything else it can be?  Describe exactly what values are accepted.

File: deepchem/models/normalizing_flows.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+    Parameters
+    ----------
+    base_distribution : tfd.Distribution
+      Probability distribution to be transformed.
+      Typically an N dimensional multivariate Gaussian.
+    flow_layers : Sequence[tfb.Bijector]
+      An iterable of bijectors that comprise the flow.
+    optimizer: Optional[tf.keras.optimizers.Optimizer]
+      An instance of Optimizer.
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:121]: Can we make it accept `dc.models.Optimizar` objects?  We should also have a `learning_rate` argument.

File: deepchem/models/normalizing_flows.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+    >> flow_layers = [
+    ..    tfb.RealNVP(
+    ..        num_masked=2,
+    ..        shift_and_log_scale_fn=tfb.real_nvp_default_template(
+    ..            hidden_layers=[8, 8]))
+    ..]
+    >> base_distribution = tfd.MultivariateNormalDiag(loc=[0., 0., 0.])
+    >> nfm = NormalizingFlowModel(base_distribution, flow_layers)
+    >> X = np.random.rand(5, 3).astype(np.float32)
+    >> nfm.build()
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:139]: This isn't needed.  `fit()` calls `build()` automatically.

**Comment by peastman** [comment]
The inheritance hierarchy you've created is NormalizingFlowModel extends NormalizingFlow, which extends tf.keras.Model.  As an alternative, how about having NormalizingFlowModel extend dc.models.KerasModel?  It would *contain* a NormalizingFlow rather than *being* a NormalizingFlow.  That would make the API much more consistent with everything else in DeepChem.  It also would provide lots more features.  Right now, these classes are missing a lot of basic features like checkpointing, logging to tensorboard, etc.

**Comment by ncfrey** [comment]
> The inheritance hierarchy you've created is NormalizingFlowModel extends NormalizingFlow, which extends tf.keras.Model. As an alternative, how about having NormalizingFlowModel extend dc.models.KerasModel? It would _contain_ a NormalizingFlow rather than _being_ a NormalizingFlow. That would make the API much more consistent with everything else in DeepChem. It also would provide lots more features. Right now, these classes are missing a lot of basic features like checkpointing, logging to tensorboard, etc.

That sounds good - I ran into an issue or two trying to inherit from `dc.models.KerasModel`, but I will work on that to take advantage of all the available features and to make this API more consistent.

**Comment by ncfrey** [comment]
`NormalizingFlow` now inherits from `tf.keras.models.Model`. `NormalizingFlowModel` inherits from `dc.models.KerasModel` and contains a `NormalizingFlow`. I will remove the commented out code that implemented the simple training loop, but otherwise I think this is ready for another review.

File: deepchem/models/losses.py [hunk 5]
```diff
@@ -126,6 +126,16 @@ def __call__(self, output, labels):
     return tf.nn.sparse_softmax_cross_entropy_with_logits(labels, output)
 
 
+class NegLogLoss(Loss):
```
Comment (peastman) [inline deepchem/models/losses.py:129]: Could you explain what this class is for?  I see several things that look like potential problems.  1. The name is confusing because it doesn't take a log.  2. It doesn't look like a loss function at all, because it completely ignores the labels.  It looks more like a regularizer?  3. Loss functions should produce a one or two dimensional (depending on the input) tensor as output with a loss for each sample or sample/task pair.  That gets multiplied by the weights from the dataset.  This only produces a single number.

File: deepchem/models/normalizing_flows.py [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+    self.model = model
+    self.flow = model.flow  # normalizing flow
+    self.loss = loss()
+    self.batch_size = batch_size
+    self.learning_rate = learning_rate
+
+    self.optimizer = optimizer(learning_rate=learning_rate)
+
+    self.built = False
+    self.build()
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:162]: Most of this conflicts with what the superclass is already doing.  For model, batch_size, learning_rate, etc. you don't even need to explicitly list arguments.  KerasModel will pull them out of `kwargs` and set appropriate fields.  Take a look at `__init__()` and `_ensure_built()` in KerasModel.

File: deepchem/models/normalizing_flows.py [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+
+    super(NormalizingFlowModel, self).__init__(
+        model=self.model, loss=self.loss, optimizer=self.optimizer, **kwargs)
+
+  def build(self):
+    """Initialize tf network."""
+    x = self.flow.distribution.sample(self.flow.distribution.batch_shape)
+    for b in reversed(self.flow.bijector.bijectors):
+      x = b.forward(x)
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:171]: Since this method is always called from the constructor, how about just moving these three lines directly into the constructor?  Then `build()` and `built` become unnecessary.

File: deepchem/models/losses.py (hunk 5 above)
Comment (ncfrey) [reply deepchem/models/losses.py:129]: Since I'm using `tfp.distributions.TransformedDistribution`, there are already methods available to compute log likelihoods of samples and the Kullback-Leibler divergence between the learned distribution and the target distribution. I will work on implementing these losses in a way that is consistent with `dc.models.losses.Loss`.

**Comment by ncfrey** [comment]
This has taken me a while to get to, but here are some more details on loss functions for normalizing flows that will hopefully clear things up a bit. This is related to the WGAN implementation mentioned in this [forum post](https://forum.deepchem.io/t/need-help-understanding-new-wgan/229/5).

Typical `dc.models.losses.Loss` functions take `outputs` and `labels` and compute the loss for each sample in a batch. For normalizing flows, the loss function is the negative log-likelihood (NLL) of observed data. The NF learns the underlying probability distribution of the training data, and tries to minimize the negative log-likelihood that the training data is drawn from that learned distribution.

Similarly, an NF can use the Kullback-Leibler (KL) divergence, which measures the difference between the true underlying distribution and the learned one. This is (I think) equivalent to doing maximum likelihood estimation on the NF model parameters, given the training data, i.e. the model parameters are fit to maximize the likelihood that the training data was drawn from the learned distribution.

Subclasses of `tfp.distributions.TransformedDistribution` have KL divergence and NLL available, so I think it makes sense to take advantage of that, rather than computing the change of variables and Jacobians within DeepChem.

I am working on implementing these loss functions in a way that follows `dc.models.gan.GAN`, where there is a `create_generator_loss` method that uses NLL. Happy to hear any comments or suggestions on if `GAN` is the right template to follow, otherwise I'll try to get the updates pushed soon for comments.

**Comment by ncfrey** [comment]
I added a `create_nll` method, similar to the `GAN` class. I think this is ready for another review!

File: deepchem/models/normalizing_flows.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+  distribution p(y).
+
+  Normalizing flows combine the advantages of autoregressive models
+  (which provide likelihood estimation but do not learn features) and
+  variational autoencoders (which learn feature representations but
+  do not provide marginal likelihoods).
+
+  """
+
+  def __init__(self, base_distribution, flow_layers, **kwargs):
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:37]: Can you add type annotations?

File: deepchem/models/normalizing_flows.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+  def __call__(self, *inputs, training=True):
+    return self.flow.bijector.forward(*inputs)
+
+
+class NormalizingFlowModel(KerasModel):
+  """A base distribution and normalizing flow for applying transformations.
+
+  A distribution implements two main operations:
+    1. Sampling from the transformed distribution.
+    2. Calculating log probabilities.
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:79]: Before you dive into the technical details, how about having a paragraph right at the top that describes what sorts of problems normalizing flows are useful for?  Imagine a confused user reading through the documentation.  Help them to figure out right away whether they should keep reading, or go on to some other class that's more likely to be useful for them.

File: deepchem/models/normalizing_flows.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+  References
+  ----------
+  .. [1] Papamakarios, George et al. "Normalizing Flows for Probabilistic Modeling and Inference." (2019). https://arxiv.org/abs/1912.02762.
+
+  """
+
+  def __init__(self, model: NormalizingFlow, **kwargs):
+    """Creates a new NormalizingFlowModel.
+
+    Parameters
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:106]: Could you add one sentence above this?  "In addition to the following arguments, this class also accepts all the keyword arguments from KerasModel."

File: deepchem/models/normalizing_flows.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    ----------
+    model: NormalizingFlow
+      An instance of NormalizingFlow.
+    loss: dc.models.losses.Loss, default NegLogLoss
+      Loss function
+    optimizer: dc.models.optimizers.Optimizer, default Adam
+      Optimizer.
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:113]: I think you removed these arguments but forgot to remove them from the docstring.

File: deepchem/models/normalizing_flows.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+
+    try:
+      import tensorflow_probability as tfp
+      tfd = tfp.distributions
+      tfb = tfp.bijectors
+    except ModuleNotFoundError:
+      raise ValueError(
+          "This class requires tensorflow-probability to be installed.")
+
+    self.model = model
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:146]: This gets done in the superclass constructor.  Just move the call up to here and then you can remove this line.

File: deepchem/models/normalizing_flows.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    self.flow = model.flow  # normalizing flow
+    """Initialize tf network."""
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:148]: Is this supposed to be a comment?

File: deepchem/models/normalizing_flows.py [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    x = self.flow.distribution.sample(self.flow.distribution.batch_shape)
+    for b in reversed(self.flow.bijector.bijectors):
+      x = b.forward(x)
```
Comment (peastman) [inline deepchem/models/normalizing_flows.py:151]: What are these lines for?  You compute `x` but never look at it.

File: docs/index.rst [hunk 15]
```diff
@@ -121,6 +121,7 @@ discussions about research, development or any general questions. If you'd like
 .. toctree::
    :glob:
    :maxdepth: 2
+
```
Comment (peastman) [inline docs/index.rst:124]: Did this file get added accidentally?

File: deepchem/models/normalizing_flows.py (hunk 14 above)
Comment (ncfrey) [reply deepchem/models/normalizing_flows.py:151]: I've seen these lines in other normalizing flow examples to "initialize the network" by applying all the transformations once, but I'm not completely clear on if that is necessary or what is happening under the hood. I'll experiment with removing these lines.

**Comment by ncfrey** [comment]
Happy to make any further changes on this, otherwise the basic API is working and it would be helpful to merge this in while I work on the tutorial notebook. 

CC @peastman @rbharath

**Review by rbharath — COMMENTED** [review]
Looks good on my end!

**Comment by rbharath** [comment]
Actually, had a quick thought. It would be good to add a save/reload test for `NormalizingFlowModel` to this or a follow-up PR. I've been having a lot of trouble in #2151 debugging saving/reloading so we should add tests for new models as we add them in.

**Comment by ncfrey** [comment]
If it's ok, I will push that to a future PR. I actually had trouble saving/reloading because of an issue between Keras and the TFP bijector layers. I'll add a comment to #2151 to track that.

**Comment by rbharath** [comment]
Ok sounds good! Once this PR looks good to @peastman, then we should be clear to merge then

**Comment by peastman** [comment]
I'm unlikely to have time to review it in the next few days.  I'll trust your judgement on it!
